docs/technical-reference/Mutation Orchestration Design.md

```diff
@@ -0,0 +1,320 @@
+- `PrepareContext(...)`: creates a `MutantControl.Member` context.
+- `RestoreContext()`: leaves the current context
+
+#### BaseFunctionOrchestrator
+`BaseFunctionOrchestrator<T>` is an abstract base class helping orchestrate methods, functions (locals and anonymous) as well as accessors. This class will convert  from expression body to block body form if any mutation requires it. This class also implements `IInstrumentCode` and provides the adequate roll back logic for the compilation phase. It inherits from `MemberDefinitionOrchestrator<T>`.
+Child classes must implement abstract methods that abstract some common operations:
+- `ParameterList(T)`: must return the `ParameterListSyntax` instance listing all the provided method’s parameters. Returns an empty list if there are none.
+- `ReturnType(T)`:  must return the `TypeSyntax` for the method’s returning type.  Should return `null` If the type cannot be determined (anonymous lambdas). One can use `RoslynHelper.VoidTypeSyntax()` if the method is void , explicitly or implicitly (constructors for example).
+- `GetBodies(T)`: must return the body for the method. Return value is a tuple with the block body as first item and the expression body as the second. At least one of them should be null
+- `SwitchToThisBodies(...)`: Must replace the method body with the provided body (either block or expression). This method is used to perform the expression to block body conversion and for roll back as well. Note that implementations assumes one of the provided bodies is null (as a function cannot have both an expression body and block statement body;
+Orchestrator’s overrides:
+- `InjectMutations(...)`: if the body is in expression form and there are mutations that have not been injected (because they require statement or block level control), converts the body to block form and inject those mutations. If the body is in block form, injects out parameters initialization to default and add an ending `return`  statement if it looks helpful.
+Finally, this class implements an injection engine:
+- `ConvertToBlockBody(...)` converts the method from expression form to block body form. Does nothing if the method has no expression body.
+-  `RemoveInstrumentation(...)`: reverts the conversion. Note that this method requires that the body contains only a `return` or expression statement.
+### Specialized orchestrators
+These orchestrators are specific for certain syntax node types.
+#### MemberAccessOrchestrator
+`MemberAccessOrchestrator<T>`is used for ‘MemberAccess’ kind of expression, which is defined as a lower level than expression. This is because while those syntax constructs inherit from `ExpressionSyntax`, they cannot be replaced by other kind of expressions, which means that Stryker cannot use a conditional operator (`?:`) to mutate them in place; it must happen at a higher hierarchical level in the expression.
```

Comment:
```suggestion
`MemberAccessOrchestrator<T>` is used for ‘MemberAccess’ kind of expression, which is defined as a lower level than expression. This is because while those syntax constructs inherit from `ExpressionSyntax`, they cannot be replaced by other kinds of expression, which means that Stryker cannot use a conditional operator (`?:`) to mutate them in place; it must happen at a higher hierarchical level in the expression.
```